### PR TITLE
parallel dbkvs ddl operations

### PR DESCRIPTION
**Pull request by clockfort**
parallelize dbkvs ddl operations and also add dbkvs ddl timeouts (even for non-parallel version)

**Goals (and why)**:
Two separate issues:
- We have an upgrade task that for some users will drop over 1000 atlas tables.
  Right now Oracle KVS clocks in at about 7 seconds per drop; the upgrade task takes over 2 hours.
  (Really, Oracle KVS is internally doing two table drops (main table, overflow table) w/ purging the segments from the tablespace and two tiny row deletes for table metadata and oracle table name remapper, so to be fair Oracle is clocking in at an extremely reasonable ~3 seconds per actual drop)

- Another customer had an issue where drop table seemed to hang; I suspect maybe because of inability to grab the necessary exclusive locks, but I was not given the requisite information nor access to debug.

**Implementation Description (bullets)**:
- Add configurables for parallelism and timeout
- driveby java8ing of surrounding code

**Concerns**:
- someone with some time on their hands could eventually move all the DBKVS code from Guava Functions to to Java ones; I elected not to rock the boat too too much in this same commit.
- This affects Oracle, Postgres, and H2 KVSs.
- I have not tested the speedup from implementing this; we could merge a benchmark for this first, but the benchmark would add a lot of time to the suite. (>100 ops for good timing numbers, p99 stats would mean doing over ten minutes of wallclock time in the fully serial version). As a result it's probably easier to benchmark myself manually, which I will do and report back.

**Priority (whenever / two weeks / yesterday)**:
whenever, pending sanity check here and some testing I was planning on instead of backporting this just writing parallism higher up into our upgrade task, so no rush on your end.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1788)

### CODE REVIEW

**Comment by clockfort** [comment]
Hey @hsaraogi , mind being the person to review this one?
I think it should obviate the need for your open #1758 and you're well-versed in DbKVS stuff anyways 😄

**Comment by hsaraogi** [comment]
@clockfort sure, will take a look. Not sure how this helps #1758 given that that issue was specifically targeted to C* tests.

File: atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/DdlConfig.java [hunk 1]
```diff
@@ -61,6 +63,16 @@ public int mutationBatchSizeBytes() {
         return 2 * 1024 * 1024;
     }
 
+    @Value.Default
+    public int parallelDdlOperationConcurrency() {
+        return 16;
```
Comment (hsaraogi) [inline atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/DdlConfig.java]: Concurrency here looks okay for the upgrade task described if there are only dropTableOperations. But imagine a case where you create a table and delete the same table concurrently, and delete operation is processed first instead. We end up with the table not being purged. Do we see that as a problem?

File: atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java [hunk 2]
```diff
@@ -1081,102 +1084,109 @@ private void fillOverflowValues(DbReadTable table,
...
-                return null;
-            }
-        });
+        truncateTables(ImmutableSet.of(tableRef));
+    }
+
+    @Override
+    public void truncateTables(Set<TableReference> tableRefs) {
+        List<Future> futures = Lists.newArrayListWithCapacity(tableRefs.size());
+        tableRefs.forEach(tableRef -> futures.add(ddlExecutor.submit(() ->
```
Comment (hsaraogi) [inline atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java]: We might be able to extract a method that takes a Collection and a lambda and does the concurrent operations and `waitOnDdlOperations`.

File: atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/DdlConfig.java (hunk 1 above)
Comment (hsaraogi) [inline atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/DdlConfig.java]: Looking deeper, I realize that only one kind of operations will be performed in parallel by the executor, so this isnt a concern anymore.

Comment (clockfort) [inline atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/DdlConfig.java]: I don't see that as a problem, the actual overall KVS calls are synchronous/blocking so if a user does the sane thing and says
createTables(foo,bar)
and then dropTables(foo,bar)
It will still always be correct.

A user introducing their own concurrency with drop and create tables simultaneously already could have had issues even without the extra parallelism here.

**Comment by clockfort** [comment]
ah, I didn't realize #1758 was for cassandra specifically

File: atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java (hunk 2 above)
Comment (clockfort) [inline atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java]: yeah, I thought about it but couldn't find a great way to deal with Set elements vs. Map Entry elements in the underlying logic that didn't look gross.

**Comment by hsaraogi** [comment]
@clockfort

File: atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java (hunk 2 above)
Comment (hsaraogi) [inline atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java]: Lets extract out for truncate and drop and leave the one with Map Entry as is.

**Comment by gsheasby** [comment]
@hsaraogi @clockfort - are we planning on getting this worked on anytime soon?

File: atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbDdlTable.java [hunk 3]
```diff
@@ -16,7 +16,7 @@
 package com.palantir.atlasdb.keyvalue.dbkvs.impl;
 
 public interface DbDdlTable {
-    void create(byte[] tableMetadta);
+    void create(byte[] tableMetadata);
```
Comment (jboreiko) [inline atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbDdlTable.java]: lol

**Comment by jboreiko** [comment]
Looks good to me.  Never quite been sure what the right mechanism is here for getting tests that drive high confidence.  I guess the best would be what @hsaraogi recommended by push over the parallel value (though I imagine you could set it lower in the test to increase the speed).

**Comment by hsaraogi** [comment]
@jboreiko I would like to see this in and file an issue for parallel ddl operation tests. We could backlog it for now I guess.

**Comment by hsaraogi** [comment]
@clockfort flagging that this PR is valuable to AtlasDB, but is blocked on:
1. non-trivial merge conflicts with @gbonik's dbkvs refactors.
2. No tests for {create,truncate,drop}Tables APIs :(

Will you be able to get to these anytime soon?

File: atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java [hunk 4]
```diff
@@ -1250,11 +1254,11 @@ public String getTablePrefix() {
         }
     }
 
-    private <T> T runDdl(TableReference tableRef, Function<DbDdlTable, T> runner) {
+    private void runDdl(TableReference tableRef, Function<DbDdlTable, Void> runner) {
```
Comment (berler) [inline atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java]: this could potentially be Consumer<DbDdlTable> instead

Comment (berler) [inline atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java]: just to avoid confusion because this might not have been so clear, `Consumer<T>` imo is a better choice than `Function<T, Void>`

**Comment by hsaraogi** [comment]
@clockfort hey, do you have any bandwidth to look at it this week?

**Comment by hsaraogi** [comment]
ping!

**Comment by clockfort** [comment]
Yep, I have bandwidth this week, I'll put in an update to this ancient PR.

File: atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java [hunk 5]
```diff
@@ -1070,104 +1072,103 @@ private void fillOverflowValues(ConnectionSupplier conns,
         });
     }
 
+    private void runOnTables(Set<TableReference> tableRefs, Consumer<DbDdlTable> runner) {
```
Comment (hsaraogi) [inline atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java]: Super nit: Move below dropTables for readability.

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsDdlBenchmarks.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+import com.palantir.atlasdb.keyvalue.api.Namespace;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.performance.backend.AtlasDbServicesConnector;
+import com.palantir.atlasdb.performance.benchmarks.table.RegeneratingTable;
+
+@State(Scope.Benchmark)
+public class KvsDdlBenchmarks {
+
+    // should be larger than the default concurrency (16)
+    private static int TABLES_PER_BATCH = 128;
```
Comment (hsaraogi) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsDdlBenchmarks.java]: Can be `final`.

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsDdlBenchmarks.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    private Namespace CREATE = Namespace.create("CREATE");
```
Comment (hsaraogi) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsDdlBenchmarks.java]: these should be `static final`

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsDdlBenchmarks.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.atlasdb.performance.benchmarks;
+
+import static org.openjdk.jmh.annotations.Level.Invocation;
```
Comment (hsaraogi) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsDdlBenchmarks.java]: Unused imports :/

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsDdlBenchmarks.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    public void resetState(AtlasDbServicesConnector connector) {
+        // drop any tables we've already created
+        kvs.dropTables(
+                kvs.getAllTableNames().stream()
+                        .filter(tableReference -> tableReference.getNamespace().equals(CREATE))
+                        .collect(Collectors.toSet()));
+
+        // ensure we have an upcoming batch that truncate and drop can use
+        kvs.createTables(
+                generateNonOverlappingTableNames();
```
Comment (hsaraogi) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsDdlBenchmarks.java]: Looks like a parameter is missing here and semicolon shouldnt be there here :/

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsDdlBenchmarks.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+        )
```
Comment (hsaraogi) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsDdlBenchmarks.java]: semi-colon :p

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsDdlBenchmarks.java [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    private int startOffset = 0;
+    private KeyValueService kvs;
+
+    @Setup
+    public void setup(AtlasDbServicesConnector connector) {
+        kvs = connector.connect().getKeyValueService();
+    }
+
+    @Setup(Level.Iteration)
+    public void resetState(AtlasDbServicesConnector connector) {
```
Comment (hsaraogi) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsDdlBenchmarks.java]: This is a bit off as benchmarks generally need the tables to be annotated with `@State(Scope.Benchmark)`. See `RegeneratingTable`/`ConsecutiveNarrowTable`.

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsDdlBenchmarks.java [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+
+import com.palantir.atlasdb.AtlasDbConstants;
+import com.palantir.atlasdb.keyvalue.api.KeyValueService;
+import com.palantir.atlasdb.keyvalue.api.Namespace;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.performance.backend.AtlasDbServicesConnector;
+import com.palantir.atlasdb.performance.benchmarks.table.RegeneratingTable;
+
+@State(Scope.Benchmark)
+public class KvsDdlBenchmarks {
```
Comment (hsaraogi) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/benchmarks/KvsDdlBenchmarks.java]: Thanks a lot for adding the benchmark. I would prefer if the benchmark could be split into a different PR that we will merge a day before this one and get real signal on the perf change.

**Comment by clockfort** [comment]
whoops, forgot about finishing the benchmark code before submitting, I think I started working on that like 2 months ago >_>

**Comment by hsaraogi** [comment]
Looks good overall, should be good to merge once the benchmark for table deletes/truncates merges.

**Comment by hsaraogi** [comment]
@clockfort Still waiting on the benchmark before merging this :P

**Comment by hsaraogi** [comment]
Closing this for now, feel free to reopen if we want to finish up!
